File: typescript/src/secp256k1.ts
import { secp256k1 } from "@noble/curves/secp256k1.js";
import { decodeHex, encodeHex } from "./encoding.js";
import { keyHoldError } from "./errors.js";
import {
  PRIVATE_KEY_LENGTH_BYTES,
  PUBLIC_KEY_LENGTH_BYTES,
} from "./constants.js";

export function publicKeyFromPrivate(privateKey: Uint8Array): Uint8Array {
  if (
    privateKey.length !== PRIVATE_KEY_LENGTH_BYTES ||
    !secp256k1.utils.isValidSecretKey(privateKey)
  )
    throw keyHoldError(
      "invalid_private_key",
      "private key is outside secp256k1 range",
    );
  return secp256k1.getPublicKey(privateKey, true);
}
export function validatePublicKeyHex(value: unknown): Uint8Array {
  const bytes = decodeHex(value, "publicKeyHex");
  if (
    bytes.length !== PUBLIC_KEY_LENGTH_BYTES ||
    (bytes[0] !== 2 && bytes[0] !== 3)
  )
    throw keyHoldError("invalid_document", "invalid compressed public key");
  if (!secp256k1.utils.isValidPublicKey(bytes, true))
    throw keyHoldError("invalid_document", "invalid secp256k1 public key");
  return bytes;
}
export function publicKeyMatches(
  privateKey: Uint8Array,
  publicKeyHex: string,
): boolean {
  return encodeHex(publicKeyFromPrivate(privateKey)) === publicKeyHex;
}

File: go/secp256k1.go
package keyhold

import (
	secp "github.com/decred/dcrd/dcrec/secp256k1/v4"
	"math/big"
)

// PublicKeyFromPrivate derives a compressed secp256k1 public key.
func PublicKeyFromPrivate(privateKey []byte) ([]byte, error) {
	if len(privateKey) != PrivateKeyLengthBytes {
		return nil, E(ErrInvalidPrivateKey, "private key must be 32 bytes", nil)
	}
	d := new(big.Int).SetBytes(privateKey)
	if d.Sign() <= 0 || d.Cmp(secp.Params().N) >= 0 {
		return nil, E(ErrInvalidPrivateKey, "private key is outside secp256k1 range", nil)
	}
	key := secp.PrivKeyFromBytes(privateKey)
	return key.PubKey().SerializeCompressed(), nil
}

// ValidatePublicKeyHex validates and decodes a compressed secp256k1 public key.
func ValidatePublicKeyHex(value string) ([]byte, error) {
	bytes, err := DecodeHex(value, "publicKeyHex")
	if err != nil {
		return nil, err
	}
	if len(bytes) != PublicKeyLengthBytes || (bytes[0] != 2 && bytes[0] != 3) {
		return nil, E(ErrInvalidDocument, "invalid compressed public key", nil)
	}
	key, err := secp.ParsePubKey(bytes)
	if err != nil || len(key.SerializeCompressed()) != PublicKeyLengthBytes || !equalBytes(key.SerializeCompressed(), bytes) {
		return nil, E(ErrInvalidDocument, "invalid secp256k1 public key", err)
	}
	return bytes, nil
}
